pkg/cosign/fulcio/fulcio.go

```diff
@@ -0,0 +1,119 @@
+package fulcio
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"crypto/rand"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/base64"
+	"errors"
+	"os"
+
+	"github.com/sigstore/fulcio/cmd/client/app"
+
+	"github.com/coreos/go-oidc"
+	httptransport "github.com/go-openapi/runtime/client"
+	"github.com/go-openapi/strfmt"
+	"github.com/sigstore/fulcio/pkg/generated/client/operations"
+	"github.com/sigstore/fulcio/pkg/generated/models"
+	"github.com/sigstore/fulcio/pkg/oauthflow"
+	"golang.org/x/oauth2"
+)
+
+const defaultFulcioAddress = "https://fulcio.sigstore.dev"
+
+func fulcioServer() string {
+	addr := os.Getenv("FULCIO_ADDRESS")
+	if addr != "" {
+		return addr
+	}
+	return defaultFulcioAddress
+}
+
+func GetCert(ctx context.Context, priv *ecdsa.PrivateKey) (string, error) {
+	fcli, err := app.GetFulcioClient(fulcioServer())
+	if err != nil {
+		return "", err
+	}
+
+	pubBytes, err := x509.MarshalPKIXPublicKey(&priv.PublicKey)
+	if err != nil {
+		return "", err
+	}
+	provider, err := oidc.NewProvider(ctx, "https://accounts.google.com")
+	if err != nil {
+		return "", err
+	}
+
+	// TODO: Switch these to be creds from the sigstore project.
+	config := oauth2.Config{
+		ClientID: "237800849078-rmntmr1b2tcu20kpid66q5dbh1vdt7aj.apps.googleusercontent.com",
+		// THIS IS NOT A SECRET - IT IS USED IN THE NATIVE/DESKTOP FLOW.
+		ClientSecret: "CkkuDoCgE2D_CCRRMyF_UIhS",
+		Endpoint:     provider.Endpoint(),
+		RedirectURL:  "http://127.0.0.1:5556/auth/google/callback",
+		Scopes:       []string{oidc.ScopeOpenID, "email"},
+	}
+
+	idToken, err := oauthflow.GetIDToken(provider, config)
+	if err != nil {
+		return "", err
+	}
+
+	var claims struct {
+		Email    string `json:"email"`
+		Verified bool   `json:"email_verified"`
+	}
+	if err := idToken.ParsedToken.Claims(&claims); err != nil {
+		return "", err
+	}
+	if !claims.Verified {
+		return "", errors.New("email not verified by identity provider")
+	}
+
+	// Sign the email address as part of the request
+	h := sha256.Sum256([]byte(claims.Email))
+	proof, err := ecdsa.SignASN1(rand.Reader, priv, h[:])
+	if err != nil {
+		return "", err
+	}
+
+	bearerAuth := httptransport.BearerToken(idToken.RawString)
+
+	params := operations.NewSigningCertParams()
+	params.SetSubmitcsr(&models.Submit{
+		Pub:   strfmt.Base64(base64.StdEncoding.EncodeToString(pubBytes)),
+		Proof: strfmt.Base64(base64.StdEncoding.EncodeToString(proof)),
+	})
+	resp, err := fcli.Operations.SigningCert(params, bearerAuth)
+	if err != nil {
+		return "", err
+	}
+	return resp.Payload.Certificate, nil
+}
+
+// This is the root in the fulcio project.
+const rootPem = `-----BEGIN CERTIFICATE-----
```

Comment:
I know this is a prototype,but it might be useful to stick the root cert into a file at some point, and pass it in via a CLI flag, that way anyone using their own certificate transparency project for development, can just pass in their own.

Reply:
second thoughts, maybe not a flag (confusing for end users).

Reply:
We need to make this configurable eventually. Opened https://github.com/sigstore/cosign/issues/87